src/providers/pkcs11_provider/key_metadata.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use super::LocalIdStore;
+use super::{KeyInfo, Pkcs11Provider};
+use crate::key_info_managers;
+use crate::key_info_managers::{KeyTriple, ManageKeyInfo};
+use log::{error, warn};
+use parsec_interface::operations::psa_key_attributes::*;
+use parsec_interface::requests::{ResponseStatus, Result};
+use std::sync::RwLock;
+
+impl Pkcs11Provider {
+    // This function returns the `RWLocks` found on the `Pkcs11Provider`
+    // in the order in which they should *always* be taken. Changing the order
+    // of locking in one method can very easily result in deadlocking.
+    pub(super) fn get_ordered_locks<'a>(
+        &'a self,
+    ) -> (
+        &'a RwLock<dyn ManageKeyInfo + Send + Sync>,
+        &'a RwLock<LocalIdStore>,
+    ) {
+        (&self.key_info_store, &self.local_ids)
+    }
```

Comment:
I've added this method to help with the lock ordering. Unfortunately I couldn't think of a way to include the session counter mutex into this